libtextsecure/message_receiver.js

```diff
@@ -667,58 +667,29 @@ MessageReceiver.prototype.extend({
   async decrypt(envelope, ciphertext) {
     let promise;
 
-    // We don't have source at this point yet (with sealed sender)
-    // This needs a massive cleanup!
-    const address = new libsignal.SignalProtocolAddress(
-      envelope.source,
-      envelope.sourceDevice
-    );
-
     const ourNumber = textsecure.storage.user.getNumber();
-    const number = address.toString().split('.')[0];
-    const options = {};
-
-    // No limit on message keys if we're communicating with our other devices
-    if (ourNumber === number) {
-      options.messageKeysLimit = false;
-    }
-
-    // Will become obsolete
-    const sessionCipher = new libsignal.SessionCipher(
-      textsecure.storage.protocol,
-      address,
-      options
-    );
-
     const me = {
       number: ourNumber,
       deviceId: parseInt(textsecure.storage.user.getDeviceId(), 10),
     };
 
-    // Will become obsolete
-    const getCurrentSessionBaseKey = async () => {
-      const record = await sessionCipher.getRecord(address.toString());
-      if (!record) {
-        return null;
-      }
-      const openSession = record.getOpenSession();
-      if (!openSession) {
-        return null;
-      }
-      const { baseKey } = openSession.indexInfo;
-      return baseKey;
-    };
+    // Envelope.source will be null on UNIDENTIFIED_SENDER
+    // Don't use it there!
+    const address = new libsignal.SignalProtocolAddress(
+      envelope.source,
+      envelope.sourceDevice
+    );
 
-    // Will become obsolete
-    const captureActiveSession = async () => {
-      this.activeSessionBaseKey = await getCurrentSessionBaseKey(sessionCipher);
-    };
+    const lokiSessionCipher = new libloki.crypto.LokiSessionCipher(
+      textsecure.storage.protocol,
+      address
+    );
 
     switch (envelope.type) {
       case textsecure.protobuf.Envelope.Type.CIPHERTEXT:
         window.log.info('message from', this.getEnvelopeId(envelope));
-        promise = captureActiveSession()
-          .then(() => sessionCipher.decryptWhisperMessage(ciphertext))
+        promise = lokiSessionCipher
+          .decryptWhisperMessage(ciphertext)
```

Comment:
Does `decryptWhisperMessage` return a promise?

Reply:
Yes it does